sycl/source/backend.cpp

```diff
@@ -94,6 +103,107 @@ __SYCL_EXPORT event make_event(pi_native_handle NativeHandle,
       std::make_shared<event_impl>(PiEvent, Context));
 }
 
+std::shared_ptr<detail::kernel_bundle_impl>
+make_kernel_bundle(pi_native_handle NativeHandle, const context &TargetContext,
```

Comment:
I've already asked, but again: why are we adding the "generic" implementation? Shouldn't it always be  backend-specific per the spec?

Reply:
From skimming through the implementation it seems that Level-Zero wouldn't be in a position to answer many of the queries you have there, so those would need to be user "input". This, again, speaks for NOT adding one-fits-all definition of make-functions.

Reply:
On the contrary, I think we shouldn't add more branching to the SYCL Runtime. We already have a dispatching mechanism, called Plugin Interface. My understanding was that SYCL runtime is a shim proxy to the PI. Since `pi_native_handle` is some sort of opaque pointer-like thing, we can easily pass our custom object like this:

```c++
namespace sycl::level_zero {
struct ZeSmth {
  // Members...
};
}
template <...> smth make_smth(T t) {
  Plugin.call<piextSmthCreateWithNativeHandle>(t);
  return smth;
}

pi_result piextSmthCreateWithNativeHandle(pi_native_handle NativeSmth) {
  auto *ZeSmthPtr = pi_cast<ZeSmth*>(NativeSmth);

  // Use ZeSmth members
}
```

> I think we should add new stuff with the possibility of controlling the ownership from the beginning.

The same way we can handle ownership by simply adding a new field to `ZeSmth`. We can pass through as much data as we need, and we don't have to add new APIs for each and every backend.

> To me this suggest that we should not define the default ones, as specializations "must be defined".

Right, but it does not say that specializations must be *explicitly* defined. In this case we have and implicit specialization, that works just fine.